src/_guides/libraries/c-interop.md

```diff
@@ -0,0 +1,142 @@
+---
+title: "C interop using FFI"
+description: "To use C code in your Dart program, use the dart:ffi library (currently in preview)."
+hw: "https://github.com/mjohnsullivan/ffi/blob/master/hello_world"
+---
+
+Dart mobile, command-line, and server apps running on the [Dart Native
+platform](/platforms/) can use the dart:ffi library to call native C APIs.
+The _ffi_ name stands for [_foreign function interface._][FFI]
+Other terms for similar functionality include _native interface_
+and _language bindings._
+
+The dart:ffi library is [in active development][ffi issue]
+and isn't complete yet.
+Although its API reference documentation hasn't been published yet,
+<!-- https://github.com/dart-lang/sdk/issues/37946 -->
+you can find its source code
+[in the SDK repo.](https://github.com/dart-lang/sdk/tree/master/sdk/lib/ffi)
+
+## Examples
+
+The following examples show how to use the dart:ffi library:
+
+| **Example** | **Description** |
+| [hello_world][] | Shows how to call a C function with no arguments and no return value. |
+| [primitives][] | Shows how to call C functions that have arguments and return values that are **ints or pointers**. Also demonstrates using **varargs**.
+| [structs][] | Shows how to use structs to pass **strings** to and from C and to handle **simple and complex C structures**. |
+| [sqllite][] | An example in the Dart SDK repo that comes with a [mini tutorial][]. |
+
+
+## Walkthrough of hello_world
+
+The [hello_world example][hello_world] has the minimum necessary code for
+calling a C library.
+
+### Files
+
+The hello_world example has the following files:
+
+| **Source file** | **Description** |
+| [hello.dart]({{ page.hw}}/hello.dart) | A Dart file that uses the `hello_world()` function from a C library. |
+| [pubspec.yaml]({{ page.hw}}/pubspec.yaml) | The usual Dart [pubspec](/tools/pub/pubspec), with a lower bounds on the SDK that's at least 2.5. |
+| [c/hello.h]({{ page.hw}}/c/hello.h) | Declares the `hello_world()` function. |
+| [c/hello.c]({{ page.hw}}/c/hello.c) | A C file that imports `hello.h` and defines the `hello_world()` function. |
+| [c/Makefile]({{ page.hw}}/c/Makefile) | A macOS-specific build file that compiles the C code into a dynamic library. |
+{:.table .table-striped }
+
+[PENDING: say something about setup.sh? It doesn't seem necessary for this example, but maybe it's needed by other examples?]
+<!-- 
+  | [setup.sh]({{ page.hw}}/setup.sh) | A macOS-specific script that sets an environment variable. [PENDING: Omit from this list? Why is it necessary? I didn't seem to need it.] |
+-->
+
+Building the C library creates two additional files:
+
+| **Generated file** | **Description** |
+| hello_world.dylib | The dynamic library loaded by the Dart app. |
+| c/hello.o | An intermediate object file. |
+{:.table .table-striped }
+
+### Building and running
+
+Here's an example of building the dynamic library and executing the Dart app:
+
+```terminal
+$ cd c
+$ make dylib
```

Comment:
Is `make dylib` a command?

Reply:
(but we could perhaps explain what it does)

Reply:
@mjohnsullivan what would you like to do?

Reply:
Taking a closer look, I see that the gcc line is just the output of the make command (i.e. what the make command is actually doing). I think it makes sense in the rendered output:

![image](https://user-images.githubusercontent.com/2164483/64459603-262e5100-d0ad-11e9-9e78-97d3e31ed2a1.png)